### PR TITLE
Support open compare page directly

### PR DESCRIPTION
**Pull request by lunny**
This PR add support to `/<owner>/<repo>/compare` web endpoint and the default branch is not a requirement now.

### CODE REVIEW

File: routers/web/repo/compare.go [hunk 1]
```diff
@@ -215,21 +215,25 @@ func ParseCompareInfo(ctx *context.Context) *CompareInfo {
...
+		infos = []string{baseRepo.DefaultBranch, baseRepo.DefaultBranch}
+	} else {
+		infos = strings.SplitN(infoPath, "...", 2)
+		if len(infos) != 2 {
+			infos = []string{baseRepo.DefaultBranch, infoPath}
+			if strings.Contains(infoPath, "..") {
+				infos = strings.SplitN(infoPath, "..", 2)
+				ci.DirectComparison = true
+				ctx.Data["PageIsComparePull"] = false
+			}
```
Comment (wxiaoguang) [inline routers/web/repo/compare.go]:
```suggestion
			if infos = strings.SplitN(infoPath, "..", 2); len(infos) == 2 {
				ci.DirectComparison = true
				ctx.Data["PageIsComparePull"] = false
			} else {
				infos = []string{baseRepo.DefaultBranch, infoPath}
			}
```

We do not need extra `strings.Contains`

File: routers/web/repo/compare.go [hunk 2]
```diff
@@ -215,21 +215,25 @@ func ParseCompareInfo(ctx *context.Context) *CompareInfo {
...
 		}
-	}
 
-	if len(infos) != 2 {
-		log.Trace("ParseCompareInfo[%d]: not enough compared branches information %s", baseRepo.ID, infos)
-		ctx.NotFound("CompareAndPullRequest", nil)
-		return nil
+		if len(infos) != 2 {
+			log.Trace("ParseCompareInfo[%d]: not enough compared branches information %s", baseRepo.ID, infos)
```
Comment (wxiaoguang) [inline routers/web/repo/compare.go]: It's impossible to get `len(infos) != 2` now, if yes, then there should be a bug. Should we use `log.Critical` here?

**Comment by lunny** [comment]
@wxiaoguang all done.

**Review by wxiaoguang — APPROVED** [review]

**Review by Gusted — APPROVED** [review]

**Review by zeripath — APPROVED** [review]

**Comment by 6543** [comment]
lgtm

**Comment by zeripath** [comment]
Make lgtm work

**Pull request merged**